src/foam/dao/MaterializedDAO.js

```diff
@@ -115,6 +125,25 @@ foam.CLASS({
         cmd.setIndex(new MaterializedDAOIndex(this));
 
         getSourceDAO().cmd(cmd);
+
+        String[] daoKeys = getObservedDAOs();
+        if (daoKeys.length != 0) {
+          for ( String daoKey : daoKeys ) {
+            DAO dao = (DAO) getX().get(daoKey);
+            var self = this;
+            if ( dao != null ) dao.listen(new Sink() {
```

Comment:
You could create a new AbstractSink instead which would save you from implementing remove(), eof() and reset().